.github/workflows/coverage.yml

```diff
@@ -12,18 +12,90 @@ on:
 
 jobs:
   coverage:
+    env:
+      badge_path: .github/badges/coverage.json
     runs-on: ubuntu-latest
+    permissions:
+      contents: write
     steps:
     - uses: actions/checkout@v4
+      with:
+        # Set branch so that can commit to it later
+        ref: ${{ github.head_ref }}
     - name: Set up Python 3.11
       uses: actions/setup-python@v5
       with:
         python-version: "3.11"
         cache: pip
-        cache-dependency-path: requirements-test-3.11.txt
+        cache-dependency-path: requirements-dev-3.11.txt
     - name: Upgrade pip
       run: python -m pip install --upgrade pip
     - name: Install test dependencies
-      run: pip install --no-dependencies -r requirements-test-3.11.txt
+      run: pip install --no-dependencies -r requirements-dev-3.11.txt
     - name: Assess coverage of unit tests
       run: pytest tests/unit --cov
+    - name: Extract total coverage percentage
+      id: cov
+      # Need to run this in own step to ensure coverage is evaluated and cast to string
+      run: echo "percentage=$( coverage report --format=total )" >> $GITHUB_OUTPUT
+    - name: Choose badge colour
+      id: design
+      env:
+        percentage: ${{ steps.cov.outputs.percentage }}
+      run: |
+        echo "colour=${{
+          env.percentage >= 90 && 'green' ||
+          env.percentage >= 70 && 'yellow' ||
+          env.percentage >= 50 && 'orange' ||
+          'red'
+        }}" >> $GITHUB_OUTPUT
+    - name: Create badges directory if necessary
+      run: mkdir -p .github/badges
+    - name: Store copy of old badge config for comparison
+      id: old_file
+      run: |
+        if [ -f $badge_path ]; then
+          var=$( cat $badge_path )
+        else
+          var=""
+        fi
+        {
+          echo "file<<EOF"
+          echo $var
+          echo EOF
+        } >> $GITHUB_OUTPUT
+    - name: Generate badge config JSON
+      run: |
+        echo "coverage = ${{ steps.cov.outputs.percentage }}%"
+        echo "colour = ${{ steps.design.outputs.colour }}"
+        {
+          echo "{"
+          echo "  \"schemaVersion\": 1,"
+          echo "  \"label\": \"Coverage\","
+          echo "  \"message\": \"${{ steps.cov.outputs.percentage }}%\","
+          echo "  \"color\": \"${{ steps.design.outputs.colour }}\""
+          echo "}"
+        } > $badge_path
+    - name: Store copy of new badge config for comparison
+      id: new_file
+      run: |
+        {
+          echo "file<<EOF"
+          echo $( cat $badge_path )
+          echo EOF
+        } >> $GITHUB_OUTPUT
```

Comment:
Can this be combined with the above step so that the file is inserted both into $badge_path and Github output in the same step? As with the other comments, not critical but maybe slightly simpler?

Reply:
Now redundant as it's the hash being saved to GITHUB_OUTPUT.